### PR TITLE
Add fake HPU mode to Habana components with dummy habana_frameworks module.

### PR DESCRIPTION
**Pull request by jmaksymczuk**
Refactor and improvements for #180

### CODE REVIEW

**Comment by jmaksymczuk** [comment]
Running bash format.sh throws errors regarding dummy modules - [attr-defined], not sure what to do about that.
Except for that the code is ready for review @kzawora-intel

File: vllm/utils.py [hunk 1]
```diff
@@ -208,10 +208,30 @@ def is_neuron() -> bool:
 
 @lru_cache(maxsize=None)
 def is_hpu() -> bool:
+    return _is_habana_frameworks_installed() or _is_built_for_hpu()
+
+
+@lru_cache(maxsize=None)
+def is_fake_hpu() -> bool:
+    return os.environ.get('VLLM_USE_FAKE_HPU', '0') != '0' or (
+        not _is_habana_frameworks_installed() and _is_built_for_hpu())
```
Comment (madamczykhabana) [inline vllm/utils.py:217]: This is a bit risky. If for whatever reason we cannot find habana_frameworks or other env issue we shouldn't fallback to CPU by default and we should fail asap unless CPU fallback was requested.

File: vllm/utils.py [hunk 2]
```diff
@@ -1088,3 +1114,69 @@ async def _run_task_with_lock(task: Callable, lock: asyncio.Lock, *args,
     """Utility function to run async task in a lock"""
     async with lock:
         return await task(*args, **kwargs)
+
+
+def _create_dummy_modules():
```
Comment (madamczykhabana) [inline vllm/utils.py:1119]: This is very brittle. Anytime someone adds a new module in a any file we'd need to remember to wrap it here. Couldn't we do it somehow differently?

File: vllm/utils.py [hunk 3]
```diff
@@ -1088,3 +1114,69 @@ async def _run_task_with_lock(task: Callable, lock: asyncio.Lock, *args,
...
+        habana_frameworks.torch.internal.bridge_config)
+
+    torch.hpu = types.ModuleType('torch.hpu')  # type: ignore
+    sys.modules['torch.hpu'] = torch.hpu
+
+    habana_frameworks.torch.core.mark_step = lambda: print(  # type: ignore
+        'calling mark_step')
+    habana_frameworks.torch.utils.internal.is_lazy = lambda: print(  # type: ignore
+        'calling is_lazy')
+    torch.hpu.synchronize = lambda: print('calling synchronize'  # type: ignore
```
Comment (madamczykhabana) [inline vllm/utils.py:1160]: How does this correspond to definitions in _migrate_to_cpu()?

File: vllm/worker/habana_model_runner.py [hunk 4]
```diff
@@ -757,13 +760,12 @@ def _prepare_prompt(
...
-                                    (self.lora_config.max_loras + 1) *
-                                    self.lora_config.max_lora_rank,
-                                    dtype=self.lora_config.lora_dtype)
+            lora_mask = torch.zeros(
+                len(seq_group_metadata_list) * max_prompt_len,
+                (self.lora_config.max_loras) * self.lora_config.max_lora_rank,
+                dtype=self.lora_config.lora_dtype)
             lora_logits_mask = torch.zeros(len(seq_group_metadata_list),
-                                           (self.lora_config.max_loras + 1) *
+                                           (self.lora_config.max_loras) *
```
Comment (madamczykhabana) [inline vllm/worker/habana_model_runner.py:768]: this looks completely unrelated and most likely this PR needs to be updated with recent changes from habana_main

File: vllm/worker/habana_worker.py [hunk 5]
```diff
@@ -138,6 +140,11 @@ def determine_num_available_blocks(self) -> Tuple[int, int]:
 
         # Execute a forward pass with dummy inputs to profile the memory usage
         # of the model.
+        if is_fake_hpu():
+            #            self.model_runner.profile_run()
```
Comment (madamczykhabana) [inline vllm/worker/habana_worker.py:144]: dead code

File: vllm/worker/habana_worker.py [hunk 6]
```diff
@@ -356,15 +364,17 @@ def init_worker_distributed_environment(
...
-            backend="hccl",
+            backend=backend,
             world_size=parallel_config.world_size,
             rank=rank,
             init_method=distributed_init_method,
         )
 
     # A small all_reduce for warmup & checking conformance.
-    dummy_tensor_hpu = torch.ones(1).to('hpu')
+    device = 'hpu' if not is_fake_hpu() else 'cpu'
```
Comment (madamczykhabana) [inline vllm/worker/habana_worker.py:376]: I've seen this snippet before. Couldn't we wrap it in a helper function? Like hpu_device_str() and move the check inside?

**Review by madamczykhabana — CHANGES_REQUESTED** [review]

File: vllm/utils.py (hunk 3 above)
Comment (jmaksymczuk) [reply vllm/utils.py]: Removed.

File: vllm/worker/habana_model_runner.py (hunk 4 above)
Comment (jmaksymczuk) [reply vllm/worker/habana_model_runner.py]: Resolved

File: vllm/worker/habana_worker.py (hunk 5 above)
Comment (jmaksymczuk) [reply vllm/worker/habana_worker.py]: Removed

File: vllm/worker/habana_worker.py (hunk 6 above)
Comment (jmaksymczuk) [reply vllm/worker/habana_worker.py]: Wrapped in a helper function

**Comment by jmaksymczuk** [comment]
Addressed most review comments, since the original PR there were changes in habana_main that cause this code to fail. Currently working on a fix.

File: vllm/utils.py (hunk 2 above)
Comment (jmaksymczuk) [reply vllm/utils.py:1131]: I've done research and asked a few people and I have not found a different way of doing it unfortunately. I'm open for suggestions but for now I have not found a more "elegant" way of doing that.

File: vllm/utils.py (hunk 1 above)
Comment (jmaksymczuk) [reply vllm/utils.py]: Changed, is_fake_hpu now only depends on VLLM_USE_FAKE_HPU flag.

**Comment by jmaksymczuk** [comment]
@madamczykhabana All review comments addressed. After merging fixed habana_main all checks pass.

File: vllm/utils.py (hunk 2 above)
Comment (madamczykhabana) [reply vllm/utils.py:1130]: What about using MagicMock?
https://stackoverflow.com/a/37126323
https://docs.python.org/3/library/unittest.mock.html

As far as I understand it should automatically mock everything in the hierarchy below. We could do it only for 'habana_frameworks'.

**Review by madamczykhabana — CHANGES_REQUESTED** [review]

File: vllm/utils.py (hunk 2 above)
Comment (jmaksymczuk) [reply vllm/utils.py:1130]: Unfortunately MagicMock doesn't solve the submodules issue but it highly improves visibility and mokes further dummy modules additions much simpler -> Changed origin dummy modules handling to MagicMock.

**Comment by jmaksymczuk** [comment]
@madamczykhabana Changed to MagicMock -> ready to merge.

File: vllm/utils.py (hunk 2 above)
Comment (madamczykhabana) [reply vllm/utils.py:1130]: Hmm... Perhaps something like this could work:
```python
builtin_import = __builtins__.__import__

def import_wrapper(name, *args, **kwargs):
        if 'habana_frameworks' in name:
                sys.modules[name] = MagicMock()
        return builtin_import(name, *args, **kwargs)

__builtins__.__import__ = import_wrapper
```

Could you please check if it works? (last thing, I promise! 😄 )

**Review by madamczykhabana — CHANGES_REQUESTED** [review]

**Comment by jmaksymczuk** [comment]
@madamczykhabana It works! I made a small refactor of dummy modules handling, all habana_frameworks submodules are now created automatically, changing it should not be required in future development. By default methods from dummy submodules do nothing, in case we need a function to be doing something (ie. return False) see: https://github.com/HabanaAI/vllm-fork/pull/250/files#diff-dab7693bd00a09e22d39aee684a7e419aa358a47c4bd20df33d44f5adf60d304R1151-R1153
PR ready to be merged.

**Review by madamczykhabana — APPROVED** [review]
LGTM
